### PR TITLE
[Merged by Bors] - chore(*): tweak priorities for linear algebra

### PR DESCRIPTION
**Pull request by sgouezel**
We make sure that the canonical path from `NonAssocSemiring` to `Ring` passes through `Semiring`,
as this is a path which is followed all the time in linear algebra where the defining semilinear map
`σ : R →+* S` depends on the `NonAssocSemiring` structure of `R` and `S` while the module
definition depends on the `Semiring` structure.

Tt is not currently possible to adjust priorities by hand (see lean4#2115). Instead, the last
declared instance is used, so we make sure that `Semiring` is declared after `NonAssocRing`, so
that `Semiring -> NonAssocSemiring` is tried before `NonAssocRing -> NonAssocSemiring`.

---

I have tried other options to streamline things, with less success. This change is both the least disruptive and the best I have found.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by sgouezel** [comment]
!bench

File: Mathlib/Algebra/Ring/Defs.lean [hunk 1]
```diff
@@ -100,9 +100,18 @@ theorem distrib_three_right [Mul R] [Add R] [RightDistribClass R] (a b c d : R)
...
-### Semirings
--/
+### Classes of semirings and rings
+
+We make sure that the canonical path from `NonAssocSemiring` to `Ring` passes through `Semiring`,
+as this is a path which is followed all the time in linear algebra where the defining semilinear map
+`σ : R →+* S` depends on the `NonAssocSemiring` structure of `R` and `S` while the module
+definition depends on the `Semiring` structure.
 
+Tt is not currently possible to adjust priorities by hand (see lean4#2115). Instead, the last
```
Comment (semorrison) [inline Mathlib/Algebra/Ring/Defs.lean:110]:
```suggestion
It is not currently possible to adjust priorities by hand (see lean4#2115). Instead, the last
```

File: Mathlib/Algebra/Ring/Defs.lean [hunk 2]
```diff
@@ -320,25 +351,6 @@ end HasDistribNeg
 ### Rings
```
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Defs.lean:351]: This heading is now a bit stale

Comment (sgouezel) [reply Mathlib/Algebra/Ring/Defs.lean:351]: It still looks useful to me: it says that now we are going to talk about properties of rings, while we were dealing with semirings above.

File: Mathlib/Tactic.lean [hunk 3]
```diff
@@ -7,6 +7,7 @@ import Mathlib.Tactic.Backtracking
 import Mathlib.Tactic.Basic
 import Mathlib.Tactic.ByContra
 import Mathlib.Tactic.Cache
+import Mathlib.Tactic.CancelDenoms
```
Comment (fpvandoorn) [inline Mathlib/Tactic.lean:10]: Is this PR somehow have a wrong configuration / base branch, that this line exists?

**Review by fpvandoorn — COMMENTED** [review]
LGTM

It's clearly an improvement overall, but it's unfortunate that there are also regressions (in MVPolynomial).

bors d+

File: Mathlib/Tactic.lean (hunk 3 above)
Comment (sgouezel) [reply Mathlib/Tactic.lean:10]: Yes, I couldn't tell why this line was not picked up by my merge with master, and I had to add it back by hand to make the linter happy. I will merge and see if bors complains.

**Pull request closed without merging**